torch_glow/src/TorchGlowTraining.cpp

```diff
@@ -0,0 +1,197 @@
+#include "glow/Optimizer/GraphOptimizer/TrainingPreparation.h"
+#include "glow/Support/Error.h"
+
+namespace glow {
+
+namespace {
+
+bool isValidSamplesForSingleInput(const Tensor &input, const Tensor &samples) {
+  return input.dims() == samples.dims().slice(1) &&
+         input.getElementType() == samples.getElementType();
+}
+
+bool isValidSamplesForSliceInput(const Tensor &input, const Tensor &samples) {
+  return input.dims().size() > 1 &&
+         input.dims().slice(1) == samples.dims().slice(1) &&
+         input.getElementType() == samples.getElementType();
+}
+
+} // namespace
+
+void TorchGlowTraining::clear() {
+  F_ = nullptr;
+  TF_ = nullptr;
+  inputPHs_.clear();
+  outputPHs_.clear();
+  selectedPH_ = nullptr;
+  bindings_.clear();
+  engine_.clear();
+}
+
+TorchGlowTraining::~TorchGlowTraining() { clear(); }
+
+llvm::Error TorchGlowTraining::init(llvm::StringRef modelFile,
+                                    std::vector<torch::jit::IValue> &inputs,
+                                    llvm::StringRef backend,
+                                    const ONNXWriterParameters &parameters,
+                                    const PyTorchLoaderSettings &settings,
+                                    const TrainingConfig &config,
+                                    RandomizeWeights mode) {
+  // Clean up all previous allocations, if any.
+  clear();
+  // Initialize execution engine.
+  engine_.setBackendName(backend);
+  // Create glow Function, only one object will be in module.
+  F_ = engine_.getModule().createFunction("torch_glow_model");
+  // Execution lambda.
+  auto setup = [&]() -> llvm::Error {
+    // Detect the proper loader.
+    if (modelFile.endswith_lower(".pt")) {
+      RETURN_IF_ERR(PyTorchFileLoader::loadPyTorchGraph(
+          modelFile.str(), inputs, *F_, inputPHs_, outputPHs_, settings, true));
+      if (mode == RandomizeWeights::AUTO) {
+        mode = RandomizeWeights::YES;
+      }
+    } else if (modelFile.endswith_lower(".onnx")) {
+      llvm::Error err = llvm::Error::success();
+      MARK_ERR_CHECKED(err);
+      ONNXModelLoader loader(modelFile.str(), {}, {}, *F_, &err);
+      RETURN_IF_ERR(err);
+      if (mode == RandomizeWeights::AUTO) {
+        mode = RandomizeWeights::NO;
+      }
+    } else {
+      RETURN_ERR("Unrecognized file extension, expect *.pt or *.onnx.");
+    }
+
+    RETURN_ERR_IF_NOT(inputPHs_.size() == 1,
+                      glow::strFormat("Only 1 input is supported, got %lu",
+                                      inputPHs_.size()));
+    RETURN_ERR_IF_NOT(outputPHs_.size() == 1,
+                      glow::strFormat("Only 1 output is supported, got %lu",
+                                      outputPHs_.size()));
+
+    // Initialization succeeded, prepare for training.
+    bindings_.allocate(engine_.getModule().getPlaceholders());
+    TensorInitializer initializer;
+    switch (mode) {
+    case RandomizeWeights::AUTO:
+      assert(false);
```

Comment:
&& "this should be converted YES or NO by this point"
or something like that

Reply:
it was, see lines 73, 81.